### ISSUE TITLE
Issue Report: Non-deterministic outputs with LMCache under concurrent inference on Qwen3 hybrid model

### ISSUE THREAD

**Issue by bingfeiz** [LMCache/LMCache#4886]
### Environment
- **Model**: Qwen3.6-27B-FP8 (Mamba + Attention hybrid)
- **vLLM**: v0.25.1 (custom image `vllm-lmcache:v0.25.1`)
- **LMCache**: v0.5.2dev50, server-client split architecture via `LMCacheMPConnector`
- **GPUs**: 2x NVIDIA L20 (Compute Capability 8.9), TP=2
- **CUDA**: 13.1 (Driver 590.48.01)
- **Concurrency**: 64

### Key configuration
**vLLM args**: `--mamba-cache-mode align --kv-cache-dtype fp8_e4m3 --enable-prefix-caching --enable-chunked-prefill --max-model-len 65536 --max-num-batched-tokens 3135 --gpu-memory-utilization 0.90 --compilation-config.mode 3 --compilation-config.cudagraph_mode PIECEWISE`
**LMCache Server args**: `--chunk-size 1568 --separate-object-groups --l1-size-gb 50 --l1-init-size-gb 8 --l2-prefetch-max-in-flight 32 --eviction-policy noop --l2-store-policy skip_l1 --prometheus-port 9020`
**L2 adapter**: `{"type":"fs_native","base_path":"/data/llm/lmcache/disk/","num_workers":32,"max_capacity_gb":800,"eviction":{"eviction_policy":"LRU","trigger_watermark":0.8,"eviction_ratio":0.2}}`
**kv-transfer-config**: `{"kv_connector":"LMCacheMPConnector","kv_connector_module_path":"lmcache.integration.vllm.lmcache_mp_connector","kv_role":"kv_both","kv_connector_extra_config":{"lmcache.mp.host":"tcp://localhost","lmcache.mp.port":9018}}`
**Sampling params**: `temperature=0, top_p=0.8, top_k=20, min_p=0.0, presence_penalty=1.5, repetition_penalty=1.15`

### Test Setup
GSM8K evaluation (1319 questions) using `gsm8k_eval_test.py`, running two consecutive rounds against the same vLLM instance with `temperature=0`, `seed=42`, `use-chat-completions`.
The evaluation script builds few-shot prompts with a **dynamic prefix** -- every N test questions share the same few-shot prefix constructed from the training set. The number of shots also acts as the group size for prefix sharing.

### Control: Baseline (no LMCache, pure vLLM)
| Shots | Round 1 Acc | Round 2 Acc | Divergence Rate |
| :--- | :--- | :--- | :--- |
| 50 | 0.957 | 0.958 | ~6% |

Even without LMCache, pure vLLM baseline shows ~6% divergence between two rounds.

### LMCache (varying configurations)
| Shots | Config | Round 1 Acc | Round 2 Acc | Divergence Rate |
| :--- | :--- | :--- | :--- | :--- |
| 48 | fp8 + noop + skip_l1 | 0.934 | 0.908 | **7.9%** |
| 50 | fp8 + noop + skip_l1 | 0.955 | 0.758 | **24.1%** |
| 50 | fp16 + noop + skip_l1 | 0.945 | 0.828 | **15.8%** |
| 50 | fp8 + LRU + default | 0.957 | 0.692 | **30.2%** |
| 64 | fp8 + noop + skip_l1 | 0.956 | 0.838 | **14.9%** |

### Observations
1. **Round 2 accuracy is consistently lower than Round 1** -- despite identical parameters (temperature=0, fixed seed). In the worst case, accuracy dropped from 0.957 to 0.692.
2. **Invalid response rate increases in Round 2** -- from near 0% up to 6.7%.
3. **fp16 ('auto') reduces but does not eliminate the divergence** -- from 24.1% down to 15.8%.
4. **Divergence rate varies with num-shots** -- 48 shows ~7.9%, 64 shows ~14.9%, 50 shows ~24.1%. Different shot counts produce different prefix lengths, which changes the amount of KV cache data stored via L2.

### Request
We hope the LMCache team can help analyze the root cause of this issue. If needed, I can provide additional logs or files for reproduction.

**Comment by zhengfeihe** [comment]
Hi~

Thanks for the report! There were some issues with hybrid models historically. Could you try the latest vLLM and LMCache? I think it should resolve this issue.

**Comment by bingfeiz** [comment]
Thank you for your reply.

We upgraded to the latest versions as suggested. Here are the details.

### Environment

- **Model**: Qwen3.6-27B-FP8 (Mamba + Attention hybrid)
- **vLLM**: 0.27.1 (image: lmcache/vllm-openai:latest)
- **LMCache**: 0.5.4 (same image)
- **GPUs**: 2x NVIDIA L20 (Compute Capability 8.9), TP=2
- **CUDA**: 13.1 (Driver 590.48.01)

### Test Method

GSM8K evaluation (1319 questions), two consecutive rounds with identical parameters:
- temperature=0, seed=42, top_p=0.8, top_k=20, min_p=0.0
- presence_penalty=1.5, repetition_penalty=1.15
- max_tokens=512, use-chat-completions, concurrency=64

### vLLM Key Args
- --mamba-cache-mode align --kv-cache-dtype fp8_e4m3 --enable-prefix-caching --enable-chunked-prefill
- --max-model-len 65536 --max-num-batched-tokens 3135

### LMCache Server Args
- --chunk-size 1568 --separate-object-groups --l1-size-gb 50 --eviction-policy noop --l2-store-policy skip_l1
- L2 adapter: fs_native, LRU eviction, 800 GB SSD

### Results

### Baseline (pure vLLM, no LMCache)

| Shots | Round 1 Acc | Round 2 Acc | Divergence |
|-------|-------------|-------------|------------|
| 50    | 0.963       | 0.956       | 2.6%       |

### LMCache (shot=50, fp8, noop+skip_l1)

| Round 1 Acc | Round 2 Acc | Divergence |
|-------------|-------------|------------|
| 0.933       | 0.041       | 93.5%      |

Round 2 has an invalid response rate of 95.1%. All outputs are garbled text, e.g.:

> "ury有意ibe1_ALIGNMENTTunc...iz...Top1...profurch..."

It looks like binary data is being misinterpreted when reading from the KV cache.

For comparison, the old version (LMCache 0.5.2dev50 + vLLM 0.25.1) with the same configuration had Round 2 accuracy drop from 0.955 to 0.758 (24.1% divergence), with a very low invalid rate. While the new versions are newer, the Round 2 outputs are almost completely corrupted.

Note: This issue has already been reported as [#4247](https://github.com/LMCache/LMCache/issues/4247)

**Comment by zhengfeihe** [comment]
@bingfeiz 

 Thanks for the detailed report! I remember that issue,  it's been around for a while and seems to affect quite a few users, with several follow-up reports. I'll try to reproduce it on my local machine today with the existing PR, and will also ping others to take a look.

**Comment by zhengfeihe** [comment]
Hi~ @bingfeiz 

 I've reproduced this locally, and I think the PR linked in #4247 (#4253) is a working fix.

This bug only affects running vLLM with FlashInfer as the attention backend. In your setup, --kv-cache-dtype fp8_e4m3 makes vLLM fall back to FlashInfer (FlashAttention doesn't support fp8 KV cache on Ada GPUs), which is why you hit it.

I think we need this fix and will push it forward, probably add a regression CI test as well.Thanks for the detailed report and  your time for helping track this down!

**Comment by bingfeiz** [comment]
Hi @zhengfeihe ,

Thank you so much for the detailed explanation.

I have a follow-up question regarding the accuracy drop (divergence) we observed in other configurations.

In our initial tests with the older version (vLLM 0.25.1 / LMCache 0.5.2dev50), we reviewed the outputs and did not observe any garbled text. Additionally, I also tried fp16, and the accuracy drop issue persists.

Since this accuracy degradation occurs even when the KV cache retrieval does not produce garbled outputs, could this be related to recurrent state synchronization in hybrid models? Or is there another known cause for this specific accuracy drop under shared-prefix requests?

Thank you again for taking the time to help. I would greatly appreciate any insights or guidance you can provide!

**Comment by zhengfeihe** [comment]
Hi~

@bingfeiz 

> Additionally, I also tried fp16, and the accuracy drop issue persists

This is also on vllm 0.25.1? There were some issue fixed for serving hybrid models on the vllm side, which might be the cause for your problem. 

Eg: https://github.com/vllm-project/vllm/issues/46453
